code/scheme/guile-r7rs/bbhw.scm

```diff
@@ -3,29 +3,23 @@
         (scheme read)
         (srfi 18))
 
-(define (valid-count? count)
-  (and (integer? count) (>= count 0)))
-
-(define (prompt-count)
-  (write-string "countdown: ")
-  (let* ((input (read-line))
-         (count (string->number input)))
-    (if (valid-count? count)
-        count
-        (begin (write-string "Invalid countdown ")
-               (write-string input)
-               (write-string ", try again...\n")
-               (prompt-count)))))
-
 (define count
-  (let ((from-cmdline (string->number (cadr (append (command-line) (list ""))))))
-    (if (valid-count? from-cmdline)
-        from-cmdline
-        (prompt-count))))
+  (do ((input (cadr (append (command-line) (list "")))
+              (begin
+                (write-string "countdown: ")
+                (read-line))))
+      ((let ((count (string->number input)))
+         (and (integer? count) (>= count 0)))
+       (string->number input))
```

Comment:
The multiple calls to `string->number` are unavoidable here as the ⟨test⟩ and ⟨expression⟩ are separate expressions in the `do` construct.